File: packages/core/src/framing.ts
// Binary frame: [u32 BE streamId][payload]. Used for PTY bytes and proxied TCP.

export namespace Frame {
  export const HEADER = 4

  export interface Decoded {
    stream: number
    payload: Uint8Array
  }

  export const encode = (stream: number, payload: Uint8Array): Uint8Array => {
    const out = new Uint8Array(HEADER + payload.length)
    new DataView(out.buffer).setUint32(0, stream)
    out.set(payload, HEADER)
    return out
  }

  export const decode = (buf: Uint8Array): Decoded => {
    if (buf.length < HEADER) throw new Error(`frame too short: ${buf.length}`)
    const stream = new DataView(buf.buffer, buf.byteOffset, buf.byteLength).getUint32(0)
    return { stream, payload: buf.subarray(HEADER) }
  }
}

File: internal/wire/framing.go
package wire

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// HeaderSize is the [u32 BE streamId] prefix on every binary frame. What follows is
// either PTY bytes for one sandbox or one proxied TCP connection's bytes — which of the
// two is decided when the stream is opened, never per frame.
const HeaderSize = 4

var ErrShortFrame = errors.New("wire: frame shorter than its header")

// Encode prefixes payload with its stream id.
func Encode(stream uint32, payload []byte) []byte {
	out := make([]byte, HeaderSize+len(payload))
	binary.BigEndian.PutUint32(out, stream)
	copy(out[HeaderSize:], payload)
	return out
}

// Decode splits a binary frame. The payload aliases buf — copy it before the read buffer
// is reused, which is what every caller that hands bytes to another goroutine must do.
func Decode(buf []byte) (stream uint32, payload []byte, err error) {
	if len(buf) < HeaderSize {
		return 0, nil, fmt.Errorf("%w: %d bytes", ErrShortFrame, len(buf))
	}
	return binary.BigEndian.Uint32(buf), buf[HeaderSize:], nil
}
